File: 78-subsets/subsets.cpp
class Solution {
public:
    void f(int i,vector<int>&v,vector<int>&nums,vector<vector<int>>&res,int n){
        if(i==n){
            res.push_back(v);
            return;
        }

        // take
        v.push_back(nums[i]);
        f(i+1,v,nums,res,n);
        v.pop_back();

        // leave
        f(i+1,v,nums,res,n);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>>res;
        int n=nums.size();
        vector<int>v;
        f(0,v,nums,res,n);

        return res;
    }
};

File: 78-subsets/subsets.java
class Solution {
    void f(int i,int n,int []nums,List<Integer>v,List<List<Integer>>ans){
        if(i==n){
            ans.add(new ArrayList<>(v));
            return;
        }

        // take
        v.add(nums[i]);
        f(i+1,n,nums,v,ans);
        v.remove(v.size()-1);

        // not take
         f(i+1,n,nums,v,ans);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>>ans=new ArrayList<>();
        int n=nums.length;
        List<Integer>v=new ArrayList<>();
        f(0,n,nums,v,ans);
        return ans;
    }
}